### PR TITLE
Tests: upgrade `test_eager_matches_sdpa_generate`

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

`test_eager_matches_sdpa_generate` has failed in our new failure reporting system ([here](https://huggingface.co/datasets/hf-internal-testing/transformers_daily_ci/raw/ccbc621b1791f7713e2d8e977a481844a4a97140/2024-10-24/ci_results_run_models_gpu/new_model_failures_with_bad_commit_grouped_by_authors.json), cc @ydshieh )

Having a look at the test, the cause for flakiness was clear: we are using random models with `generate`, and tiny perturbations can result in a different sampled token, causing generation to go in a different direction and ultimately failing the check (eager generate == sdpa generate).

This PR:
1. Moves the test to `GenerationTesterMixin`, as it calls `generate`
2. Adds logic to handle the flakiness and do the correct check. If the generation is different, check the logits in the first different token. Despite resulting in a different token, the logits should be nearly identical -- if they are, the test pass.
3. Remove most overwrites, which only existed to handle flakiness through `is_flaky()`.

________________________________

The following test commands were run:
1. `RUN_SLOW=1 py.test tests/models/ -k test_eager_matches_sdpa_generate` ✅ 
2. `RUN_SLOW=1 py.test tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_eager_matches_sdpa_generate --flake-finder --flake-runs 500` ✅

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -2046,6 +2047,81 @@ def test_inherits_generation_mixin(self):
         for model_class in self.all_generative_model_classes:
             self.assertTrue("GenerationMixin" in str(model_class.__bases__))
 
+    @require_torch_sdpa
```
Comment (gante) [inline tests/generation/test_utils.py:2050]: (this is mostly copy-paste, going to comment the sections that are changed)

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -2046,6 +2047,81 @@ def test_inherits_generation_mixin(self):
...
+
+                # Just test that a large cache works as expected
+                generate_kwargs = {
+                    "max_new_tokens": max_new_tokens,
+                    "do_sample": False,
+                    "return_dict_in_generate": True,
+                    "output_scores": True,
+                }
+                res_eager = model_eager.generate(**inputs_dict, **generate_kwargs)
+                res_sdpa = model_sdpa.generate(**inputs_dict, **generate_kwargs)
```
Comment (gante) [inline tests/generation/test_utils.py:2098]: Uses dictionaries -> more compact

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -2046,6 +2047,81 @@ def test_inherits_generation_mixin(self):
...
+                            continue
+                        first_mismatch_idx = batch_matches.int().argmin()  # gets the index of the first False
+                        first_mismatch_idx -= input_length  # scores doesn't include data regarding input tokens
+                        sdpa_first_mismatch_scores = res_sdpa.scores[first_mismatch_idx][batch_idx]
+                        eager_first_mismatch_scores = res_eager.scores[first_mismatch_idx][batch_idx]
+                        has_matching_scores = torch.allclose(
+                            sdpa_first_mismatch_scores, eager_first_mismatch_scores, rtol=1e-3, atol=1e-3
+                        )
+                        if not has_matching_scores:
+                            break
```
Comment (gante) [inline tests/generation/test_utils.py:2121]: flakiness handling as explained in the PR header

File: tests/generation/test_utils.py [hunk 4]
```diff
@@ -2046,6 +2047,81 @@ def test_inherits_generation_mixin(self):
...
+                    torch_dtype=torch.float16,
+                    low_cpu_mem_usage=True,
+                ).to(torch_device)
+
+                model_eager = model_class.from_pretrained(
+                    tmpdirname,
+                    torch_dtype=torch.float16,
+                    low_cpu_mem_usage=True,
+                    attn_implementation="eager",
+                ).to(torch_device)
```
Comment (ArthurZucker) [inline tests/generation/test_utils.py:2088]: with #34282 we won't have to init 2 models. 
Also this is memory hungry

**Review by ArthurZucker — APPROVED** [review]
Nice!

**Review by ydshieh — APPROVED** [review]
This is sooooo great! ❤️

File: tests/generation/test_utils.py (hunk 4 above)
Comment (gante) [reply tests/generation/test_utils.py:2088]: TBH we can do it one model at a time, going to change the test.

After Flex attention becomes the norm, we probably won't need this test
